### PR TITLE
Add continue option to crawler cli tool

### PR DESCRIPTION
**Pull request by lamaks**
#### Description of changes

Add continue option to crawler cli tool
Refactor store of a scan metadata

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Added relevant unit test for your changes. (`yarn test`)
- [ ] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/cli/src/runner/crawler-command-runner.ts [hunk 1]
```diff
@@ -46,6 +51,18 @@ export class CrawlerCommandRunner implements CommandRunner {
         await this.generateSummaryReports(scanArguments, scanResult, startDate, startDateNumber, endDate, endDateNumber);
     }
 
+    private canRun(scanArguments: ScanArguments): boolean {
+        if (this.filesystem.existsSync(scanArguments.output) && scanArguments.restart === false && scanArguments.continue === false) {
```
Comment (waabid) [inline packages/cli/src/runner/crawler-command-runner.ts:55]: Is there a test for this?

File: packages/cli/src/cli.ts [hunk 2]
```diff
@@ -101,6 +107,10 @@ function readScanArguments(): ScanArguments {
                 if (isEmpty(args.url)) {
                     throw new Error('The --url option is required.');
                 }
+
+                if (args.restart === true && args.continue === true) {
```
Comment (waabid) [inline packages/cli/src/cli.ts:111]: Seems like we don't have tests for this or the cli-client-runner changes. Is this intentional/necessary?

File: packages/cli/src/runner/crawler-command-runner.ts (hunk 1 above)
Comment (lamaks) [reply packages/cli/src/runner/crawler-command-runner.ts]: added test

File: packages/cli/src/cli.ts (hunk 2 above)
Comment (lamaks) [reply packages/cli/src/cli.ts:111]: will add later

File: packages/cli/src/report/report-disk-writer.ts [hunk 3]
```diff
@@ -6,59 +6,68 @@ import filenamifyUrl from 'filenamify-url';
 import * as fs from 'fs';
 import { injectable } from 'inversify';
 import { isEmpty } from 'lodash';
+import normalizePath from 'normalize-path';
 import * as path from 'path';
 import { ReportFormats } from './report-formats';
 
+// tslint:disable: no-parameter-reassignment
```
Comment (waabid) [inline packages/cli/src/report/report-disk-writer.ts:13]: Why is this required?

File: packages/cli/src/runner/crawler-command-runner.spec.ts [hunk 4]
```diff
@@ -3,79 +3,148 @@
...
-    const testUrl = 'http://www.bing.com';
-    const testInput: ScanArguments = { url: testUrl, output: '/users/xyz' };
-    // tslint:disable-next-line: mocha-no-side-effect-code
+
     beforeEach(() => {
+        testInput = { url: testUrl, output: './dir' };
+        crawlerOption = {
+            baseUrl: testInput.url,
+            localOutputDir: testInput.output,
+            existingUrls: undefined,
```
Comment (waabid) [inline packages/cli/src/runner/crawler-command-runner.spec.ts:46]: why do these have to be defined? why not just cast the object instead?

Comment (lamaks) [reply packages/cli/src/runner/crawler-command-runner.spec.ts:46]: example?

File: packages/cli/src/report/report-disk-writer.ts (hunk 3 above)
Comment (lamaks) [reply packages/cli/src/report/report-disk-writer.ts]: left over

File: packages/cli/src/report/report-disk-writer.spec.ts [hunk 5]
```diff
@@ -31,37 +32,67 @@ describe('ReportDiskWriter', () => {
         testSubject = new ReportDiskWriter(fsMock.object, pathMock.object);
     });
 
+    afterEach(() => {
+        fsMock.verifyAll();
```
Comment (waabid) [inline packages/cli/src/report/report-disk-writer.spec.ts:36]: Not sure what the pattern is here but IIRC the pattern in the web project is to avoid verify in afterEach method.

File: packages/cli/src/runner/crawler-command-runner.spec.ts (hunk 4 above)
Comment (waabid) [reply packages/cli/src/runner/crawler-command-runner.spec.ts:46]:
```

crawlerOption = {
    baseUrl: testInput.url,
    localOutputDir: testInput.output,
} as CrawlerRunOptions;
```
should do the trick.

Comment (lamaks) [reply packages/cli/src/runner/crawler-command-runner.spec.ts:46]: does not work, need to have explicit undefined

File: packages/cli/src/report/report-disk-writer.spec.ts (hunk 5 above)
Comment (lamaks) [reply packages/cli/src/report/report-disk-writer.spec.ts:36]: the pattern here is to have common mocks setup to verify so they are not missed or skip. Had several cases when this helped. It will not break the ability verify locally in test  It will also verify any setup set in a test or skip if it is not set in test.

File: packages/cli/src/runner/crawler-command-runner.spec.ts (hunk 4 above)
Comment (waabid) [reply packages/cli/src/runner/crawler-command-runner.spec.ts:46]: Hm, I just tried it on my end, and it seemed to work I didn't even need the `as` casting. Hmmmm...

**Review by waabid — APPROVED** [review]
LGTM, minor comments.
